### PR TITLE
MDEV-22939 Avoid server crash on ALTER TABLE ENGINE INNODB after DISCARD TABLESPACE

### PR DESCRIPTION
**Pull request by shih-che**
Currently MariaDB 10.3 and 10.4 will crash when executing ALTER TABLE {tablename} ENGINE INNODB against a discarded table. (More context in https://jira.mariadb.org/browse/MDEV-22939).

This PR is an attempt to fix such crash and keep on-disk files consistent with in-memory structure after the alter table query. A testcase is added for verification purpose.

### CODE REVIEW

File: mysql-test/suite/innodb/t/alter_inplace_after_discard.test [hunk 1]
```diff
@@ -0,0 +1,7 @@
+--source include/have_innodb.inc
+CREATE TABLE t (a INT) ENGINE=INNODB;
+ALTER TABLE t DISCARD TABLESPACE;
+ALTER TABLE t ENGINE INNODB;
+ALTER TABLE t ENGINE INNODB;
+RENAME table t to t1;
+DROP TABLE t1;
```
Comment (dr-m) [inline mysql-test/suite/innodb/t/alter_inplace_after_discard.test:7]: The test could be named `alter_discard`. Please do not omit the newline character at the end of the last line.

File: storage/innobase/row/row0merge.cc [hunk 2]
```diff
@@ -4336,7 +4336,8 @@ row_merge_rename_tables_dict(
 	renamed is a single-table tablespace, which must be implicitly
 	renamed along with the table. */
 	if (err == DB_SUCCESS
-	    && old_table->space_id) {
+	    && old_table->space_id
+		&& old_table->space != NULL) {
```
Comment (dr-m) [inline storage/innobase/row/row0merge.cc:4340]: I would prefer an implicit comparison to `NULL` whenever possible. This should fit in a single line:
```c++
 	if (err == DB_SUCCESS && old_table->space_id && old_table->space) {
```

**Comment by dr-m** [comment]
Hi! I could also try to fix this in my way (first, by checking why this does not fail in the 10.2 and 10.5 series), but to avoid duplicating work, I would like to know whether you are going to address my review comments.

**Comment by shih-che** [comment]
Hi Marko,

I will address the review comments, but currently I somehow lost my 2FA credentials and are struggling to recover them(I’m very sorry about this). Once I’m done with that I will be able to make the changes.
And if you want to fix it in your way, I’m totally fine with discarding my changes as I did not quite work on this after submitting the pull request. After all I think the bug needs a more thorough fix whereas my solution looks more like a workaround.

**Comment by dr-m** [comment]
Hi @shih-che, thank you and good luck recovering your GitHub account.

I may take over this then. There are also other bugs in this area, such as the newly filed [MDEV-23705](https://jira.mariadb.org/browse/MDEV-23705).

**Comment by dr-m** [comment]
Closed by 3eb81136e1c087af3fef38ae89e3357f4c306f8d and 78efa1093076307946873f322b45a14d1ed5ffb0.

**Pull request closed without merging**